apps/test-app/frontend/src/components/tree-widget/stateless-v2/TreeActions.ts

```diff
@@ -0,0 +1,292 @@
+  rootNodes: Array<PresentationTreeNode> | undefined;
+  isLoading: boolean;
+}
+
+/** @internal */
+export class TreeActions {
+  private _loader: IHierarchyLoader;
+  private _currentModel: TreeModel = { idToNode: {}, parentChildMap: new Map(), rootNode: { id: undefined, nodeData: undefined } };
+  private _disposed = new Subject<void>();
+
+  constructor(private _updater: (actionOrValue: TreeState | ((initialState: TreeState) => TreeState)) => void) {
+    this._loader = new NoopHierarchyLoader();
+  }
+
+  private getModel(model: TreeModel, actionOrModel: ((model: Draft<TreeModel>) => void) | TreeModel) {
+    if (typeof actionOrModel === "function") {
+      return produce(model, actionOrModel);
+    }
+    return actionOrModel;
+  }
+
+  private updateTreeState(action: (state: TreeState) => TreeState) {
+    this._updater(action);
+  }
+
+  private updateTreeModel(actionOrModel: ((model: Draft<TreeModel>) => void) | TreeModel) {
+    const newModel = this.getModel(this._currentModel, actionOrModel);
+    if (this._currentModel === newModel) {
+      return;
+    }
+
+    this._currentModel = newModel;
+    const rootNodes = this._currentModel.parentChildMap.get(undefined) !== undefined ? generateTreeStructure(undefined, this._currentModel) : undefined;
+    this._updater({
+      model: this._currentModel,
+      rootNodes,
+      isLoading: false,
+    });
+  }
+
+  private loadNodes(parentId: string | undefined) {
+    const parentNode = parentId ? this._currentModel.idToNode[parentId] : this._currentModel.rootNode;
+    if (parentNode.id !== undefined && !isTreeModelHierarchyNode(parentNode)) {
+      return;
+    }
+
+    this._loader
+      .getNodes(parentNode, (node) => !!node.autoExpand)
+      .pipe(
+        takeUntil(this._disposed),
+        reduce<LoadedHierarchyPart, TreeModel>(
+          (treeModel, hierarchyPart) => {
+            addNodesToModel(treeModel, hierarchyPart);
+            return treeModel;
+          },
+          {
+            idToNode: {},
+            parentChildMap: new Map(),
+            rootNode: { id: undefined, nodeData: undefined },
+          },
+        ),
+      )
+      .subscribe({
+        next: (loadedHierarchy) => {
+          this.updateTreeModel((model) => {
+            addHierarchyPart(model, parentNode.id, loadedHierarchy);
+            const node = parentNode.id ? model.idToNode[parentNode.id] : undefined;
+            if (node && isTreeModelHierarchyNode(node)) {
+              node.isLoading = false;
+            }
+          });
+        },
+      });
+  }
+
+  public dispose() {
+    this._disposed.next();
+  }
+
+  public getNode(nodeId: string): TreeModelHierarchyNode | undefined {
+    const node = this._currentModel.idToNode[nodeId];
+    return node && isTreeModelHierarchyNode(node) ? node : undefined;
+  }
+
+  public setHierarchyProvider(provider?: HierarchyProvider) {
+    this._loader = provider ? new HierarchyLoader(provider) : new NoopHierarchyLoader();
```

Comment:
Should we do the following?
1. dispose
2. set loader
3. reload

It seems strange that we could have a load in progress and change loader somewhere in the middle, producing weird state.

Reply:
All these steps are done in `useEffect` inside `useTree`.